src/core/Fingerprint.cpp

```diff
@@ -0,0 +1,240 @@
+    m_sFingerprintPresent                  = *PFINGERPRINTPRESENT;
+}
+
+static void fingerprintUnlockCallback(std::shared_ptr<CTimer> self, void* data) {
+    g_pHyprlock->unlock();
+}
+
+void CFingerprint::start() {
+    static auto* const PENABLEFINGERPRINT = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:enable_fingerprint");
+    const bool         ENABLEFINGERPRINT  = **PENABLEFINGERPRINT;
+    if (!ENABLEFINGERPRINT) {
+        return;
+    }
+    std::thread([this]() {
+        // For grace or SIGUSR1 unlocks
+        if (g_pHyprlock->isUnlocked())
+            return;
+
+        setupDBUS();
+
+        if (m_bAuthenticated)
+            g_pHyprlock->addTimer(std::chrono::milliseconds(1), fingerprintUnlockCallback, nullptr);
+    }).detach();
+}
+
+std::optional<std::string> CFingerprint::getLastMessage() {
+    return m_sDBUSState.message.empty() ? std::nullopt : std::optional(m_sDBUSState.message);
+}
+
+void CFingerprint::terminate() {
+    if (m_sDBUSState.connection)
+        m_sDBUSState.connection->leaveEventLoop();
+}
+
+bool CFingerprint::setupDBUS() {
+    m_sDBUSState.connection = sdbus::createSystemBusConnection();
+
+    registerSleepHandler();
+
+    if (m_sDBUSState.sleeping) {
+        m_sDBUSState.connection->enterEventLoop();
+    } else if (startVerify()) {
+        m_sDBUSState.message = m_sFingerprintReady;
+        g_pHyprlock->enqueueForceUpdateTimers();
+        m_sDBUSState.connection->enterEventLoop();
+    }
+
+    if (m_sDBUSState.abort)
+        return true;
+
+    releaseDevice();
+    return true;
+}
+
+void CFingerprint::registerSleepHandler() {
+    m_sDBUSState.login = sdbus::createProxy(*m_sDBUSState.connection.get(), "org.freedesktop.login1", "/org/freedesktop/login1");
+
+    m_sDBUSState.sleeping = m_sDBUSState.login->getProperty("PreparingForSleep").onInterface(LOGIN_MANAGER);
+    m_sDBUSState.login->uponSignal("PrepareForSleep").onInterface(LOGIN_MANAGER).call([this](bool start) {
+        Debug::log(LOG, "prepare for sleep");
+        if (start) {
+            m_sDBUSState.sleeping = true;
+            stopVerify();
+            m_sDBUSState.inhibitLock.reset();
+        } else {
+            m_sDBUSState.sleeping = false;
+            inhibitSleep();
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+            if (startVerify()) {
+                m_sDBUSState.message = m_sFingerprintReady;
+                g_pHyprlock->enqueueForceUpdateTimers();
+            }
+        }
+    });
+    m_sDBUSState.login->finishRegistration();
+    if (!m_sDBUSState.sleeping)
+        inhibitSleep();
+}
+
+void CFingerprint::inhibitSleep() {
+    m_sDBUSState.login->callMethod("Inhibit")
+        .onInterface(LOGIN_MANAGER)
+        .withArguments("sleep", "hyprlock", "Release fingerprint before sleep", "delay")
+        .storeResultsTo(m_sDBUSState.inhibitLock);
+}
```

Comment:
So this is used to make sure we call `stopVerify` before we are going to sleep right?

Reply:
Right, updated the inhibit reason.